tests/src/test-load-geomodel.cpp

```diff
@@ -0,0 +1,258 @@
+
+#include <geogram/basic/line_stream.h>
+
+#include <ringmesh/geomodel/geomodel.h>
+#include <ringmesh/geomodel/geomodel_geological_entity.h>
+#include <ringmesh/geomodel/geomodel_validity.h>
+#include <ringmesh/io/io.h>
+
+/*!
+ * @file Test GeoModel building
+ * @author Arnaud Botella
+ */
+
+using namespace RINGMesh;
+
+template< index_t DIMENSION >
+void load_input_geomodel( GeoModel< DIMENSION >& geomodel, const std::string& file )
+{
+    bool loaded_model_is_valid = geomodel_load( geomodel,
+        ringmesh_test_data_path + file );
+    if( !loaded_model_is_valid )
+    {
+        throw RINGMeshException( "RINGMesh Test",
+            "Failed when loading model " + geomodel.name()
+                + ": the loaded model is not valid." );
+    }
+}
+
+template< index_t DIMENSION >
+void throw_error( const GeoModel< DIMENSION >& geomodel, const std::string& entity )
+{
+    throw RINGMeshException( "RINGMesh Test", "Failed when loading model ",
+        geomodel.name(), ": the loaded model as not the correct number of ", entity );
+}
+
+void get_line( GEO::LineInput& in )
+{
+    in.get_line();
+    in.get_fields();
+}
+
+template< index_t DIMENSION >
+void check_geomodel(
+    const GeoModel< DIMENSION >& geomodel,
+    const std::string& result );
+
+template< >
+void check_geomodel( const GeoModel< 2 >& geomodel, const std::string& result )
+{
+    std::string info { ringmesh_test_data_path + "load/" + result };
+    GEO::LineInput in( info );
+    if( !in.OK() )
+    {
+        throw RINGMeshException( "TEST", "Failed to load file: ", info );
+
+    }
+    get_line( in );
+    index_t nb_corners = in.field_as_uint( 1 );
+    if( geomodel.nb_corners() != nb_corners )
+    {
+        throw_error( geomodel, "corners" );
+    }
+    get_line( in );
+    index_t nb_lines = in.field_as_uint( 1 );
+    if( geomodel.nb_lines() != nb_lines )
+    {
+        throw_error( geomodel, "lines" );
+    }
+    get_line( in );
+    index_t nb_surfaces = in.field_as_uint( 1 );
+    if( geomodel.nb_surfaces() != nb_surfaces )
+    {
+        throw_error( geomodel, "surfaces" );
+    }
+    get_line( in );
+    index_t nb_vertices = in.field_as_uint( 1 );
+    if( geomodel.mesh.vertices.nb() != nb_vertices )
+    {
+        throw_error( geomodel, "vertices" );
+    }
+    get_line( in );
+    index_t nb_polygons = in.field_as_uint( 1 );
+    if( geomodel.mesh.polygons.nb() != nb_polygons )
+    {
+        throw_error( geomodel, "polygons" );
+    }
+    while( !in.eof() && in.get_line() )
+    {
+        in.get_fields();
+        std::string geol_type = in.field( 0 );
+        index_t nb_geol_entities = in.field_as_uint( 1 );
+        if( geomodel.nb_geological_entities( GeologicalEntityType { geol_type } )
+            != nb_geol_entities )
+        {
+            throw_error( geomodel, geol_type );
+        }
+    }
+}
+
+template< >
+void check_geomodel( const GeoModel< 3 >& geomodel, const std::string& result )
```

Comment:
Avoid code duplication between the function for GeoModel2D and this one for GeoModel3D. You can define a template function for common parts.